mesonbuild/mtest.py

```diff
@@ -283,22 +283,38 @@ def run_single_test(self, test):
             if ('MALLOC_PERTURB_' not in test_env or not test_env['MALLOC_PERTURB_']) and not self.options.benchmark:
                 test_env['MALLOC_PERTURB_'] = str(random.randint(1, 255))
 
-            setsid = None
             stdout = None
             stderr = None
             if not self.options.verbose:
                 stdout = subprocess.PIPE
                 stderr = subprocess.PIPE if self.options and self.options.split else subprocess.STDOUT
 
-            if not is_windows():
-                setsid = os.setsid
+            # Let gdb handle ^C instead of us
+            if test_opts.gdb:
+                previous_sigint_handler = signal.getsignal(signal.SIGINT)
+                # Make the meson executable ignore SIGINT while gdb is running.
+                signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+            def preexec_fn():
+                if test_opts.gdb:
+                    # Restore the SIGINT handler for the child process to
+                    # ensure it can handle it.
+                    signal.signal(signal.SIGINT, signal.SIG_DFL)
+                else:
+                    # We don't want setsid() in gdb because gdb needs the
+                    # terminal in order to handle ^C and not show tcsetpgrp()
+                    # errors avoid not being able to use the terminal.
+                    os.setsid()
+
+            if is_windows():
+                preexec_fn = None
```

Comment:
__[Flake8]__

[F811] redefinition of unused 'preexec_fn' from line 298


<!-- Comment by [SideCI](https://sideci.com) -->

Reply:
That was intentional, but arguably it's a bit obscure... Let's do it another way.